### ISSUE TITLE
How to avoid overfitting?

### ISSUE THREAD

**Issue by MadeInGer** [sagr4019/ResearchProject#6]
There exist methods to protect the system from overfitting. This issue provides an overview without any claim of completeness.

**Cross-Validation**
Offers information how well the generalization on independent data works (effectiveness of the model). 

Cross-Validation methods:
- **Holdout Method**: A part from the training data is removed and used later to get the error estimation → maybe underfitting is given, because the training data has been reduced

- **K-Fold Cross Validation**: data is divided into k subsets (sections/folds), repeats the holdout method k times. Each time a subset is used as test data or validation data, the other k-1 subsets are used as training data, error estimation is computed over all k trials, k=5 or k=10 is preferred, but not fixed.

- **Stratified K-Fold Cross Validation**: Divides the data such that the distribution of classes are held constant for each k folds, reduces the variance of the performance estimate

- **Leave-One-Out Cross Validation**: K-Fold cross validation with k equal to n (where n = number of elements)

**Early Stopping**
Training is stopped before overfitting starts (once the performance has stopped improving on a validation set)

**Pruning**
Used to optimize, simplify and shorten decision trees.

**Regularization** is a (mathematical) technique to avoid overfitting within our model. It adds a penalty to the model to handle with high complexity and increases the generalization of the model. The penalty has the most impact for high regression coefficients or if there are many coefficients that are not null.

Regularization Methods
The mean-squared error cost function is used along the examples:
![selection_218](https://user-images.githubusercontent.com/30792417/38832879-70e2634c-41b3-11e8-9ddf-3f40ae96b7f0.png)

- **Ridge Regression** (also known as **L2 regularization**): As the regularization (bordered red) the beta coefficients are squared and summed.
![selection_217](https://user-images.githubusercontent.com/30792417/38832829-4e789740-41b3-11e8-91ea-7b367d3ba9be.png)

- **Lasso Regression** (also known as **L1 regularization**):As the regularization (bordered red) the beta coefficients are summed (absolute value). 
![selection_219](https://user-images.githubusercontent.com/30792417/38832889-7a1640be-41b3-11e8-871f-3f3702050d69.png)

Both L1 and L2 are using Lambda as hyper-parameter (in short: chooses the regularization strength). Difference between L1 and L2 regularization: L1 penalizes smaller weights equally as larger weights. L2 penalizes larger weights a lot more.

- **Elastic Net Regression** (Ridge + Lasso): Includes squared and absolute values in the cost function.

- **Dropout**
Neurons selected randomly are ignored during training (dropped-out). Other neurons need to step in and handle the missing but required neurons to make predictions. At the end our network has the ability to better generalize and less overfitting. So dropout removed width from the network.

- **DropConnect**
Instead of randomly ignore neurons during training (dropout), DropConnect randomly ignores (set to zero) subsets of weights within the network.

- **Stochastic Depth**
Randomly skips entire layers during training. So stochastic depth removes depth from the network.

- **Data augmentation**
Increase the number of training data by changing the data (e. g. rotate/resize images) and adding it to the training data.

- **Ensemble Methods**
Train multiple models for same task, merges them to get a better predictive performance

**Comment by jan-christiansen** [comment]
Please, don't use issue to capture results. Move the text to an article in the wiki.

**Comment by MadeInGer** [comment]
This content has been merged to the [overfitting](https://github.com/sagr4019/ResearchProject/wiki/Overfitting) article.
